pkg/cgroups/fsscan/fsscan.go

```diff
@@ -0,0 +1,199 @@
+	// matching the pod id did not existed in the fs.
+	//
+	// Callers need to serialize concurrent access to this function on their own.
+	FindContainerPath(podID uuid.UUID, containerID string) (string, error)
+}
+
+func New() FsScanner {
+	return &fsScannerState{}
+}
+
+type fsScannerState struct {
+	knownPodDirs []string
+	root         string
+	rootErr      error
+}
+
+// We 've seen some cases, where the container cgroup file is not in a
+// directory that mathces the pod id. To handle these cases, do a full scan of
+// the root to find a directory with the container id.
+func fsFullContainerScan(root string, containerID string) string {
+	found := errors.New("Found")
+	retPath := ""
+	// NB: there might be more efficient ways of searching the file-system than WalkDir, but
+	// use that for now and we can improve later as needed.
+	filepath.WalkDir(root, func(path string, dentry fs.DirEntry, err error) error {
+		if err != nil || !dentry.IsDir() {
+			return nil
+		}
+		base := filepath.Base(path)
+		if strings.Contains(base, containerID) {
+			retPath = path
+			return found
+		}
+		return nil
+	})
+	return retPath
+}
+
+// FindContainer implements FindContainer method from FsScanner
+func (fs *fsScannerState) FindContainerPath(podID uuid.UUID, containerID string) (string, error) {
+
+	// first, check the known (cached) locations
+	for _, loc := range fs.knownPodDirs {
+		podDir, containerDir := findPodAndContainerDirectory(loc, podID, containerID)
+		if podDir == "" {
+			continue
+		}
+
+		if containerDir != "" {
+			return filepath.Join(podDir, containerDir), nil
+		}
+
+		// found the pod dir, but not the container dir. Return an error
+		return "", fmt.Errorf("found pod dir=%s but failed to find container for id=%s", podDir, containerID)
+	}
+
+	if fs.root == "" && fs.rootErr == nil {
+		fs.root, fs.rootErr = cgroups.HostCgroupRoot()
+		if fs.rootErr != nil {
+			logger.GetLogger().WithError(fs.rootErr).Warnf("failed to retrieve host cgroup root")
+		}
+	}
+	if fs.rootErr != nil {
+		return "", fmt.Errorf("no cgroup root")
+	}
+
+	podPath := fsFindPodPath(fs.root, podID)
+	if podPath == "" {
+		contPath := fsFullContainerScan(fs.root, containerID)
+		if contPath == "" {
+			return "", fmt.Errorf("no directory found that matches container-id '%s'", containerID)
+		}
+		return contPath, ErrContainerPathWithoutMatchingPodID
+	}
+	podDir := filepath.Dir(podPath)
+	// found a new pod directory, added it to the cached locations
+	fs.knownPodDirs = append(fs.knownPodDirs, podDir)
+	logger.GetLogger().Infof("adding %s to cgroup pod directories", podDir)
+	containerDir := findContainerDirectory(podPath, containerID)
+	if containerDir == "" {
+		return "", fmt.Errorf("found pod dir=%s but failed to find container for id=%s", podPath, containerID)
+	}
+	return filepath.Join(podPath, containerDir), nil
+}
+
+func podDirMatcher(podID uuid.UUID) func(p string) bool {
+	s1 := podID.String()
+	s2 := strings.Replace(s1, "-", "_", 4)
```

Comment:
Why do you use 4 here?

Reply:
s1 is "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx"
so there are only 4 "-". I've added a comment.